stl/inc/flat_map

```diff
@@ -0,0 +1,300 @@
+#pragma once
+
+#include <algorithm>
+#include <initializer_list>
+#include <compare>
+#include <concepts>
+#include <cstdint>
+#include <memory>
+#include <ranges>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include <xmemory>
+
+_STD_BEGIN
+
+template <class Key, class T, class Compare = less<Key>, class KeyContainer = vector<Key>, class MappedContainer = vector<T>> requires same_as<Key, typename KeyContainer::value_type> && same_as<T, typename MappedContainer::value_type>
+class flat_map;
+
+struct sorted_unique_t { explicit sorted_unique_t() = default; };
+inline constexpr sorted_unique_t sorted_unique {};
+
+template <class _Alloc, class _Key_container, class _Mapped_container>
+concept _Valid_Allocator_for_flat_map = _STD uses_allocator_v<_Key_container, _Alloc> && _STD uses_allocator_v<_Mapped_container, _Alloc>;
+
+template <class _Key, class _T, class _Key_compare>
+struct _Flat_value_compare {
+    struct type{
+        public:
+        bool operator () (_STD pair<const _Key&, const _T&> _X, _STD pair<const _Key&, const _T&> _Y) const {
+            return _Key_compare_for_val(_X.first, _Y.first);
+        }
+
+        type(_Key_compare _Comp) : _Key_compare_for_val(_Comp) {}
+
+        private:
+        _Key_compare _Key_compare_for_val;
+    };
+};
+
+template <class _Key_container, class _Mapped_container>
+struct _Flat_Container {
+    struct type {
+        _Key_container keys;
+        _Mapped_container values;
+    };
+};
+
+// Implementation
+
+template <_STD random_access_iterator _Key_iterator_t, _STD random_access_iterator _Mapped_iterator_t>
+class _Flat_map_iterator_Impl{
+    public:
+    class type{

```

Comment:
I think we don't have to define an inner type to do the real definitions..

Reply:
> I think we don't have to define an inner type to do the real definitions..

While this is correct, I think a nested non-template class can ensure that no user-defined type can be an associated class of the iterator, which possibly saves compiler throughput and prevents ADL-hijacking.

Reply:
I found that nested iterator types can actually be ADL-proof ([Godbolt link](https://godbolt.org/z/WdTW5e7h9)).

Reply:
@frederick-vs-ja  Does that mean we can define it as `flat_map<>::iterator` instead of using an external class `_Flat_map_iterator<>::type`?

Reply:
> @frederick-vs-ja Does that mean we can define it as `flat_map<>::iterator` instead of using an external class `_Flat_map_iterator<>::type`?

This is not I meant... I still think `_Flat_map_iterator<>::type` is better.

It's OK to directly define the iterator as a nested class in `flat_map`, but doing so will make the iterator type depend on the comparator type, which is less than ideal.